File: docmanage_mobile/lib/api/document_api.dart
import 'http_service.dart';

class DocumentApi {
  // =========================================================
  // DOCUMENTS
  // =========================================================

  static Future getDocuments() async {
    return HttpService.request(
      '/documents/count',
      method: 'GET',
    );
  }

  static Future getPersonalDocCount() async {
    return HttpService.request(
      '/documents/my/count',
      method: 'GET',
    );
  }

  static Future getDocumentList(
    Map<String, dynamic> params,
  ) async {
    return HttpService.request(
      '/documents/list',
      method: 'GET',
    );
  }

  static Future getPersonalDocumentList() async {
    return HttpService.request(
      '/documents/my',
      method: 'GET',
    );
  }

  static Future getDocument(
    String id,
  ) async {
    return HttpService.request(
      '/documents/$id',
      method: 'GET',
    );
  }

  // =========================================================
  // UPLOAD
  // =========================================================

  static Future uploadDocument({
    required String filePath,
    required String categoryId,
    required String departmentId,
  }) async {
    return HttpService.uploadMultipart(
      '/documents/upload',
      filePath: filePath,
      fileField: 'file',
      fields: {
        'categoryId': categoryId,
        'departmentId': departmentId,
      },
    );
  }

  static Future uploadPersonalDocument({
    required String filePath,
  }) async {
    return HttpService.uploadMultipart(
      '/documents/personal/upload',
      filePath: filePath,
      fileField: 'file',
    );
  }

  static Future uploadNewVersion({
    required String documentId,
    required String filePath,
  }) async {
    return HttpService.uploadMultipart(
      '/documents/$documentId/version',
      filePath: filePath,
      fileField: 'file',
    );
  }

  // =========================================================
  // DELETE / DOWNLOAD
  // =========================================================

  static Future deleteDocument(
    String id,
  ) async {
    return HttpService.request(
      '/documents/$id',
      method: 'DELETE',
    );
  }

  static Future downloadDocument(
    String id,
  ) async {
    return HttpService.request(
      '/documents/$id/download',
      method: 'GET',
    );
  }

  static Future updateDocumentStatus(
    String id,
    String statusName,
  ) async {
    return HttpService.request(
      '/documents/$id/status',
      method: 'PUT',
      body: {
        'statusName': statusName,
      },
    );
  }

  // =========================================================
  // PREVIEW
  // =========================================================

  static Future previewDocument(
    String id,
  ) async {
    return HttpService.request(
      '/documents/$id/preview',
      method: 'POST',
    );
  }

  static Future getPreviewPdf(
    String id,
  ) async {
    return HttpService.request(
      '/documents/export',
      method: 'POST',
      body: {
        'documentID': id,
      },
    );
  }

  static Future previewDocumentVersion({
    required String documentId,
    required String versionId,
  }) async {
    return HttpService.request(
      '/documents/$documentId/versions/$versionId/file',
      method: 'GET',
    );
  }

  // =========================================================
  // AI SUMMARY
  // =========================================================

  static Future getAISummary(
    String id,
  ) async {
    return HttpService.request(
      '/documents/$id/summary',
      method: 'GET',
    );
  }

  static Future generateAISummary(
    String id,
  ) async {
    return HttpService.request(
      '/documents/$id/generate-summary',
      method: 'POST',
    );
  }

  // =========================================================
  // EXPORT
  // =========================================================

  static Future exportPDF(
    String id,
  ) async {
    return HttpService.request(
      '/documents/export',
      method: 'POST',
      body: {
        'documentID': id,
      },
    );
  }

  static Future exportDOCX(
    String id,
  ) async {
    return HttpService.request(
      '/documents/export-docx',
      method: 'POST',
      body: {
        'documentID': id,
      },
    );
  }

  static Future exportXLSX(
    String id,
  ) async {
    return HttpService.request(
      '/documents/export-xlsx',
      method: 'POST',
      body: {
        'documentID': id,
      },
    );
  }

  // =========================================================
  // VERSION HISTORY
  // =========================================================

  static Future getVersionList(
    String id,
  ) async {
    return HttpService.request(
      '/documents/$id/versions',
      method: 'GET',
    );
  }

  static Future makeVersionCurrent(
    String documentId,
    int versionNum,
  ) async {
    return HttpService.request(
      '/documents/$documentId/versions',
      method: 'PUT',
      body: {
        'versionNum': versionNum,
      },
    );
  }

  static Future updateDocumentVersion(
    String id,
    int versionNum,
  ) async {
    return makeVersionCurrent(
      id,
      versionNum,
    );
  }
}

File: frontend/src/api/documentAPI.js
import { http } from './http.js';

export async function getDocumentsList(params={}){
    const cleanParams = Object.fromEntries(Object.entries(params).filter(([_, v]) => v !== null && v !== undefined && v !== ''));

    try {
        const queryString = new URLSearchParams(cleanParams).toString();
        const url = queryString ? `/documents/list?${queryString}` : '/documents/list';
        return await http (url);
    } catch (err) {
        return {
            success: false,
            message: err.message
        };
    }
}

export async function getDocuments(){
    try {
        return await http ('/documents/count');
    } catch (err) {
        return {
            success: false,
            message: err.message
        };
    }
}

export async function getDocument(id){
    try {
        
        return await http (`/documents/${id}`);
    } catch (err) {
        return {
            success: false,
            message: err.message
        };
    }
}

export async function uploadDocument(formData) {
    try{

        return await http ('/documents/upload',{
            method: "POST",
            body: formData
        });

    } catch (err) {
        return {
            success: false,
            message: err.message
        };
    }
}

export async function uploadNewVersion(documentID, formData) {
    try{
        return await http(`/documents/${documentID}/version`, {
            method: "POST",
            body: formData
        });
    } catch(err) {
        return {
            success: false,
            message: err.message
        };
    }
}

export async function deleteDocuments(id) {
    try{
        return await http (`/documents/${id}`, {
            method: "DELETE",
        });
    } catch (err) {
        return {
            success: false,
            message: err.message
        };
    }
}

export async function renameDocument(id, documentName) {
    try {
        return await http(`/documents/${id}/rename`, {
            method: "PUT",
            body: JSON.stringify({ documentName }),
        });
    } catch (err) {
        return { success: false, message: err.message };
    }
}

export async function exportPDF(documentID, options = {}){
     try{
        return http('/documents/export',{
            method:"POST",
            body:JSON.stringify({
                documentID
            }),
            ...options
        });

    }catch(err){
        return {
            success:false,
            message:err.message
        };

    }
}

export async function exportDocx(documentID, options = {}){
     try{
        return http('/documents/export-docx',{
            method:"POST",
            body:JSON.stringify({
                documentID
            }),
            ...options
        });

    }catch(err){
        return {
            success:false,
            message:err.message
        };

    }
}

export async function exportXlsx(documentID, options = {}){
     try{
        return http('/documents/export-xlsx',{
            method:"POST",
            body:JSON.stringify({
                documentID
            }),
            ...options
        });

    }catch(err){
        return {
            success:false,
            message:err.message
        };

    }
}

export async function updateDocumentStatus(id,statusName){
    try {
        return await http(`/documents/${id}/status`, {
            method: "PUT",
            body:JSON.stringify({
                statusName: statusName
            })
        });
    } catch (err) {
        return {
            success:false,
            message:err.message
        };
    }
}

export async function updateDocumentVersion(id,versionNum){
    try {
        return await http(`/documents/${id}/versions`, {
            method: "PUT",
            body:JSON.stringify({
                versionNum
            })
        });
    } catch (err) {
        return {
            success:false,
            message:err.message
        };
    }
}

export async function getVersionList(documentID){
    try{
        return await http(`/documents/${documentID}/versions`);
    } catch (err) {
        return {
            success: false,
            message: err.message
        };
    }
    
}

export async function aiService(documentID){
    try {
        return await http(`/documents/${documentID}/summary`);
    }catch (err) {
        return {
            success: false,
            message: err.message
        };
    }
}

export async function generateAISummary(documentID){
    try {
        return await http(`/documents/${documentID}/generate-summary`, {
            method: "POST"
        });
    } catch (err) {
        return {
            success:false,
            message:err.message
        };
    }
}

export async function previewDocument(documentID) {
     console.log("CALL PREVIEW API:", documentID);
     
    try{

        return await http(`/documents/${documentID}/preview`, {
            method: "POST"
        });
    } catch (err) {
        return {
            success:false,
            message:err.message
        };
    }
}

export async function toggleFavorite(documentID) {
    try {
        return await http(`/documents/${documentID}/favorite`, {
            method: "POST"
        });
    } catch (err) {
        return { success: false, message: err.message };
    }
}

export async function getFavorites() {
    try {
        return await http('/documents/favorites');
    } catch (err) {
        return { success: false, message: err.message };
    }
}

export async function isFavorite(documentID) {
    try {
        return await http(`/documents/${documentID}/is-favorite`);
    } catch (err) {
        return { success: false, message: err.message };
    }
}